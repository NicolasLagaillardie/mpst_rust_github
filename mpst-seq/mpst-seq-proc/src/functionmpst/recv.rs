use quote::{format_ident, quote};
use syn::parse::{Parse, ParseStream};
use syn::{Result, Token};

#[derive(Debug)]
pub struct RecvAuxSimpleMacroInput {
    session: proc_macro2::TokenStream,
    role: syn::Ident,
    exclusion: syn::LitInt,
}

impl Parse for RecvAuxSimpleMacroInput {
    fn parse(input: ParseStream) -> Result<Self> {
        let role = syn::Ident::parse(input)?;
        <Token![,]>::parse(input)?;
        let exclusion = syn::LitInt::parse(input)?;
        <Token![,]>::parse(input)?;
        let content;
        let _parentheses = syn::parenthesized!(content in input);
        let session = proc_macro2::TokenStream::parse(&content)?;

        Ok(RecvAuxSimpleMacroInput {
            session,
            role,
            exclusion,
        })
    }
}

impl Into<proc_macro2::TokenStream> for RecvAuxSimpleMacroInput {
    fn into(self) -> proc_macro2::TokenStream {
        self.expand()
    }
}

impl RecvAuxSimpleMacroInput {
    fn expand(&self) -> proc_macro2::TokenStream {
        let session = self.session.clone();
        let role = format_ident!("{}", self.role);

        let mut recv_vec = Vec::new();
        for i in 1..3 {
            if i == (self.exclusion).base10_parse::<u64>().unwrap() {
                recv_vec.push(format_ident!("session{}", i));
            }
        }

        let mut new_sessions = Vec::new();
        let mut all_sessions = Vec::new();
        for i in 1..3 {
            if i == (self.exclusion).base10_parse::<u64>().unwrap() {
                all_sessions.push(format_ident!("session{}", i));
                new_sessions.push(format_ident!("new_session"));
            } else {
                all_sessions.push(format_ident!("session{}", i));
                new_sessions.push(format_ident!("temp_session.session{}", i));
            }
        }

        quote! {
            || -> Result<_, Box<dyn std::error::Error>> {

                let temp_session = #session;

                let (v, new_session) = crate::binary::recv::recv( temp_session.#( #recv_vec )* )?;

                let new_stack = {
                    fn temp<R>(r: #role<R>) -> R
                    where
                        R: crate::role::Role,
                    {
                        let (here, there) = <R as crate::role::Role>::new();
                        r.sender.send(there).unwrap_or(());
                        here
                    }
                    temp(temp_session.stack)
                };

                Ok((
                    v,
                    crate::sessionmpst::SessionMpst {
                        #( #all_sessions : #new_sessions , )*
                        stack: new_stack,
                        name: temp_session.name,
                    }
                ))
            }
        }
    }
}
