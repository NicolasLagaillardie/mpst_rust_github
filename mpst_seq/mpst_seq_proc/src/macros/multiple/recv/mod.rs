pub mod create_recv_http_session;
pub mod create_recv_mpst_all_session;
pub mod create_recv_mpst_session;
pub mod recv_mpst;
