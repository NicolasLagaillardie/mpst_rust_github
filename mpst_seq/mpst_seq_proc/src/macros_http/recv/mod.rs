pub mod create_recv_http_session;
