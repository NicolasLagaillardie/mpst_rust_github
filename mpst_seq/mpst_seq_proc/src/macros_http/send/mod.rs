pub mod create_send_http_session;
