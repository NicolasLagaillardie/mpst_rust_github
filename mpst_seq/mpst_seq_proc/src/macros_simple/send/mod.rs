pub mod create_send_mpst_session;
