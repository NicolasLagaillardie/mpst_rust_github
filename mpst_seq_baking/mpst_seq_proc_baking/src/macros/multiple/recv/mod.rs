pub mod create_recv_http_session;
pub mod create_recv_mpst_session;
pub mod recv_mpst;
