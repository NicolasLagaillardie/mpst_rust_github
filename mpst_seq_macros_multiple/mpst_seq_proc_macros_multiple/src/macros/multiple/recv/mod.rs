pub mod create_recv_mpst_session;
pub mod recv_mpst;
