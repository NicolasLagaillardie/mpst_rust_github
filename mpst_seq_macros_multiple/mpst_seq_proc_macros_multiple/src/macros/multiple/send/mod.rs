pub mod create_send_check_cancel;
pub mod create_send_mpst_cancel;
pub mod create_send_mpst_session;
pub mod send_cancel;
pub mod send_mpst;
