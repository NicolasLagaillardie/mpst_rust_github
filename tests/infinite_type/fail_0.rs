use mpstthree::binary::struct_trait::end::End;
use mpstthree::binary::struct_trait::recv::Recv;
use mpstthree::binary::struct_trait::send::Send;

use std::marker;

// Bug with the constraint checker.
#[allow(dead_code)]
enum CalcOp2<N: marker::Send> {
    More(Send<i64, Recv<i64, NiceCalcServer2<N>>>),
    More2(Recv<i64, Send<i64, NiceCalcServer2<N>>>),
    Stop(Send<i64, End>),
}

#[allow(dead_code)]
type NiceCalcServer2<N> = Recv<CalcOp2<N>, End>;

fn main() {}
