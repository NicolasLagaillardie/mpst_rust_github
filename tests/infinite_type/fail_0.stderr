error[E0275]: overflow evaluating the requirement `mpstthree::binary::struct_trait::send::Send<i64, mpstthree::binary::struct_trait::recv::Recv<i64, mpstthree::binary::struct_trait::recv::Recv<CalcOp2<N>, mpstthree::binary::struct_trait::end::End>>>: std::marker::Send`
  --> tests/infinite_type/fail_0.rs:10:10
   |
10 |     More(Send<i64, Recv<i64, NiceCalcServer2<N>>>),
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: required because it appears within the type `CalcOp2<N>`
  --> tests/infinite_type/fail_0.rs:9:6
   |
9  | enum CalcOp2<N: marker::Send> {
   |      ^^^^^^^
   = note: required because of the requirements on the impl of `Session` for `mpstthree::binary::struct_trait::recv::Recv<CalcOp2<N>, mpstthree::binary::struct_trait::end::End>`
